coderd/unhanger/detector.go

```diff
@@ -0,0 +1,314 @@
+	err error
+}
+
+// Error implements error.
+func (e *acquireLockError) Error() string {
+	return "acquire lock: " + e.err.Error()
+}
+
+// Unwrap implements xerrors.Wrapper.
+func (e *acquireLockError) Unwrap() error {
+	return e.err
+}
+
+// Detector automatically detects hung provisioner jobs, sends messages into the
+// build log and terminates them as failed.
+type Detector struct {
+	ctx    context.Context
+	db     database.Store
+	pubsub database.Pubsub
+	log    slog.Logger
+	tick   <-chan time.Time
+	stats  chan<- Stats
+	done   chan struct{}
+}
+
+// Stats contains statistics about the last run of the detector.
+type Stats struct {
+	// HungJobIDs contains the IDs of all jobs that were detected as hung and
+	// terminated.
+	HungJobIDs []uuid.UUID
+	// Error is the fatal error that occurred during the last run of the
+	// detector, if any. Error may be set to AcquireLockError if the detector
+	// failed to acquire a lock.
+	Error error
+}
+
+// New returns a new hang detector.
+func New(ctx context.Context, db database.Store, pubsub database.Pubsub, log slog.Logger, tick <-chan time.Time) *Detector {
+	le := &Detector{
+		//nolint:gocritic // Hang detector has a limited set of permissions.
+		ctx:    dbauthz.AsHangDetector(ctx),
+		db:     db,
+		pubsub: pubsub,
+		tick:   tick,
+		log:    log,
+		done:   make(chan struct{}),
+	}
+	return le
+}
+
+// WithStatsChannel will cause Executor to push a RunStats to ch after
+// every tick. This push is blocking, so if ch is not read, the detector will
+// hang. This should only be used in tests.
+func (d *Detector) WithStatsChannel(ch chan<- Stats) *Detector {
+	d.stats = ch
+	return d
+}
+
+// Start will cause the detector to detect and unhang provisioner jobs on every
+// tick from its channel. It will stop when its context is Done, or when its
+// channel is closed.
+//
+// Start should only be called once.
+func (d *Detector) Start() {
+	go func() {
+		defer close(d.done)
+		for {
+			select {
+			case <-d.ctx.Done():
+				return
+			case t, ok := <-d.tick:
+				if !ok {
+					return
+				}
+				stats := d.run(t)
+				if stats.Error != nil && !xerrors.As(stats.Error, &acquireLockError{}) {
+					d.log.Warn(d.ctx, "error running workspace build hang detector once", slog.Error(stats.Error))
+				}
+				if len(stats.HungJobIDs) != 0 {
+					d.log.Warn(d.ctx, "detected (and terminated) hung provisioner jobs", slog.F("job_ids", stats.HungJobIDs))
+				}
+				if d.stats != nil {
+					select {
+					case <-d.ctx.Done():
+						return
+					case d.stats <- stats:
+					}
+				}
+			}
+		}
+	}()
+}
+
+// Wait will block until the detector is stopped.
+func (d *Detector) Wait() {
+	<-d.done
+}
+
+func (d *Detector) run(t time.Time) Stats {
+	ctx, cancel := context.WithTimeout(d.ctx, 5*time.Minute)
+	defer cancel()
+
+	stats := Stats{
+		HungJobIDs: []uuid.UUID{},
+		Error:      nil,
+	}
+
+	err := d.db.InTx(func(db database.Store) error {
+		err := db.AcquireLock(ctx, database.LockIDHangDetector)
```

Comment:
Should this be `TryAcquireLock`? The comment seems to imply the behavior of that instead of `AcquireLock`.

Reply:
good catch!